README.md
<!-- 
This README describes the package. If you publish this package to pub.dev,
this README's contents appear on the landing page for your package.

For information about how to write a good package README, see the guide for
[writing package pages](https://dart.dev/guides/libraries/writing-package-pages). 

For general information about developing packages, see the Dart guide for
[creating packages](https://dart.dev/guides/libraries/create-library-packages)
and the Flutter guide for
[developing packages and plugins](https://flutter.dev/developing-packages). 
-->

A lightweight [Tap Payment](https://www.tap.company/) SDK for flutter.

## Features

It utilizes [Flutter Webview](https://pub.dev/packages/webview_flutter), to complete the payment process... "It's a Simple Spell But Quite Unbreakable".


## Usage

After adding the package to your `pubspec.yaml`, make sure to import it


```dart
import 'package:flutter_tap_payment/flutter_tap_payment.dart';
```


You can checkout a simple illustration in the `/example` folder. 


CHANGELOG.md
## 0.2.0

- Upgraded to latest version of Flutter
- All packages updated too


test/flutter_tap_payment_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:flutter_tap_payment/flutter_tap_payment.dart';

void main() {
  // test('adds one to input values', () {
  //   final calculator = Calculator();
  //   expect(calculator.addOne(2), 3);
  //   expect(calculator.addOne(-7), -6);
  //   expect(calculator.addOne(0), 1);
  // });
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:flutter_tap_payment/flutter_tap_payment.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Tap Payment',
      theme: ThemeData(
        colorScheme: ColorScheme.fromSeed(seedColor: Colors.deepPurple),
        useMaterial3: true,
      ),
      home: const MyHomePage(title: 'Flutter Tap Payment Example'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({super.key, required this.title});
  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
        appBar: AppBar(
          title: Text(widget.title),
        ),
        body: Center(
          child: TextButton(
              onPressed: () => {
                    Navigator.of(context).push(
                      MaterialPageRoute(
                        builder: (BuildContext context) => TapPayment(
                            apiKey: "YOUR_API_KEY",
                            redirectUrl: "http://your_website.com/redirect_url",
                            postUrl: "http://your_website.com/post_url",
                            paymentData: const {
                              "amount": 10,
                              "currency": "OMR",
                              "threeDSecure": true,
                              "save_card": false,
                              "description": "Test Description",
                              "statement_descriptor": "Sample",
                              "metadata": {"udf1": "test 1", "udf2": "test 2"},
                              "reference": {
                                "transaction": "txn_0001",
                                "order": "ord_0001"
                              },
                              "receipt": {"email": false, "sms": true},
                              "customer": {
                                "first_name": "test",
                                "middle_name": "test",
                                "last_name": "test",
                                "email": "test@test.com",
                                "phone": {
                                  "country_code": "965",
                                  "number": "50000000"
                                }
                              },
                              // "merchant": {"id": ""},
                              "source": {"id": "src_card"},
                              // "destinations": {
                              //   "destination": [
                              //     {"id": "480593777", "amount": 2, "currency": "KWD"},
                              //     {"id": "486374777", "amount": 3, "currency": "KWD"}
                              //   ]
                              // }
                            },
                            onSuccess: (Map params) async {
                              print("onSuccess: $params");
                            },
                            onError: (error) {
                              print("onError: $error");
                            }),
                      ),
                    )
                  },
              child: const Text("Make Payment")),
        ));
  }
}


lib/src/TapServices.dart
// ignore_for_file: file_names

import 'dart:convert';

import 'package:flutter/material.dart';
import 'package:http/http.dart' as http;

class TapServices {
  final String apiKey;
  final Map paymentData;
  String basePath = "https://api.tap.company/";
  String version = "v2";

  TapServices({required this.apiKey, required this.paymentData});

  sendPayment() async {
    String domain = "$basePath$version/charges";
    try {
      var response = await http
          .post(Uri.parse(domain), body: jsonEncode(paymentData), headers: {
        "Authorization": "Bearer $apiKey",
        'Accept': 'application/json',
        'Content-Type': 'application/json; charset=UTF-8',
      });
      var body = json.decode(response.body);
      if (response.statusCode == 200) {
        debugPrint(body);
        return {'error': false, 'message': body};
      } else {
        debugPrint(body);
        return {
          'error': true,
          'message': "${body["errors"]?[0]?["description"]}"
        };
      }
    } catch (e) {
      debugPrint("$e");
      return {
        'error': true,
        'message': "Unable to proceed, check your internet connection."
      };
    }
  }

  confirmPayment(tapId) async {
    String domain = "$basePath$version/charges/$tapId";
    try {
      var response = await http.get(Uri.parse(domain), headers: {
        "content-type": "application/json",
        'Authorization': 'Bearer $apiKey'
      });

      final body = jsonDecode(response.body);
      if (response.statusCode == 200) {
        return {'error': false, 'message': "Confirmed", 'data': body};
      } else {
        return {
          'error': true,
          'message': "Payment inconclusive.",
          'data': body
        };
      }
    } catch (e) {
      return {'error': true, 'message': e, 'exception': true, 'data': null};
    }
  }
}


lib/src/errors/network_error.dart
import 'package:flutter/material.dart';

class NetworkError extends StatelessWidget {
  final Function loadData;
  final String message;
  final bool isSmall;

  const NetworkError(
      {super.key, required this.loadData, required this.message, this.isSmall = false});

  @override
  Widget build(BuildContext context) {
    return Center(
      child: Column(
        mainAxisAlignment: MainAxisAlignment.center,
        children: [
          Image.asset(
            "lib/src/assets/img/cloud_state.png",
            package: "flutter_tap_payment",
            height: 120,
          ),
          SizedBox(
            height: isSmall ? 20 : 40,
          ),
          Row(
            mainAxisAlignment: MainAxisAlignment.center,
            children: [
              Text(message,
                  style: const TextStyle(
                      fontSize: 14,
                      color: Color(0xFF272727),
                      fontWeight: FontWeight.w400)),
              const SizedBox(
                width: 5,
              ),
              InkWell(
                onTap: () => loadData(),
                child: const Text("Tap to retry",
                    style: TextStyle(
                        fontSize: 14,
                        color: Colors.black,
                        fontWeight: FontWeight.w800)),
              ),
            ],
          ),
        ],
      ),
    );
  }
}


lib/src/screens/complete_payment.dart
import 'package:flutter/material.dart';
import 'package:flutter_tap_payment/src/errors/network_error.dart';
import 'package:flutter_spinkit/flutter_spinkit.dart';

import '../TapServices.dart';

class CompletePayment extends StatefulWidget {
  final Function onSuccess, onError;
  final TapServices services;
  final String url;
  const CompletePayment({
    Key? key,
    required this.onSuccess,
    required this.onError,
    required this.services,
    required this.url,
  }) : super(key: key);

  @override
  // ignore: library_private_types_in_public_api
  _CompletePaymentState createState() => _CompletePaymentState();
}

class _CompletePaymentState extends State<CompletePayment> {
  bool loading = true;
  bool loadingError = false;

  String getMessage(data) {
    String message = "";
    switch (data['status']) {
      case "CAPTURED":
        message = "The transaction completed successfully";
        break;
      case "ABANDONED":
        message = "The transaction has been abandoned";
        break;
      case "CANCELLED":
        message = "The transaction has been cancelled";
        break;
      case "FAILED":
        message = "The transaction has failed";
        break;
      case "DECLINED":
        message = "The transaction has been declined";
        break;
      case "RESTRICTED":
        message = "The transaction is restricted";
        break;
      case "VOID":
        message = "The transaction is voided";
        break;
      case "TIMEDOUT":
        message = "The transaction is timedout";
        break;
      case "UNKNOWN":
        message = "The transaction is unknown";
        break;
      default:
        message = "The transaction cannot be completed";
    }
    return message;
  }

  complete() async {
    final uri = Uri.parse(widget.url);
    final tapID = uri.queryParameters['tap_id'];
    if (tapID != null) {
      setState(() {
        loading = true;
        loadingError = false;
      });

      Map resp = await widget.services.confirmPayment(tapID);
      if (resp['error'] == false) {
        if (resp['data']?['status'] == "CAPTURED") {
          Map data = resp['data'];
          data['message'] = getMessage(resp['data']);
          await widget.onSuccess(data);
          setState(() {
            loading = false;
            loadingError = false;
          });
          // ignore: use_build_context_synchronously
          Navigator.pop(context);
        } else {
          Map data = resp['data'];
          data['message'] = getMessage(resp['data']);
          widget.onError(data);
          setState(() {
            loading = false;
            loadingError = false;
          });
          // ignore: use_build_context_synchronously
          Navigator.pop(context);
        }
      } else {
        if (resp['exception'] != null && resp['exception'] == true) {
          widget.onError({"message": resp['message']});
          setState(() {
            loading = false;
            loadingError = true;
          });
        } else {
          await widget.onError(resp['data']);
          // ignore: use_build_context_synchronously
          Navigator.of(context).pop();
        }
      }
      //return NavigationDecision.prevent;
    } else {
      Navigator.of(context).pop();
    }
  }

  @override
  void initState() {
    super.initState();
    complete();
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Container(
        child: loading
            ? const Column(
                children: [
                  Expanded(
                    child: Center(
                      child: SpinKitFadingCube(
                        color: Color(0xFFEB920D),
                        size: 30.0,
                      ),
                    ),
                  ),
                ],
              )
            : loadingError
                ? Column(
                    children: [
                      Expanded(
                        child: Center(
                          child: NetworkError(
                              loadData: complete,
                              message: "Something went wrong,"),
                        ),
                      ),
                    ],
                  )
                : const Center(
                    child: Text("Payment Completed"),
                  ),
      ),
    );
  }
}


lib/flutter_tap_payment.dart
library flutter_tap_payment;

import 'dart:core';
import 'package:flutter/material.dart';
import 'package:flutter_tap_payment/src/screens/complete_payment.dart';
import 'package:flutter_spinkit/flutter_spinkit.dart';
import 'package:webview_flutter/webview_flutter.dart';
// Import for Android features.
import 'package:webview_flutter_android/webview_flutter_android.dart';
// Import for iOS features.
import 'package:webview_flutter_wkwebview/webview_flutter_wkwebview.dart';

import 'src/TapServices.dart';
import 'src/errors/network_error.dart';

class TapPayment extends StatefulWidget {
  final Function onSuccess, onError;
  final String apiKey, redirectUrl, postUrl;
  final Map paymentData;

  const TapPayment({
    Key? key,
    required this.onSuccess,
    required this.onError,
    //
    required this.apiKey,
    required this.redirectUrl,
    required this.postUrl,
    required this.paymentData,
  }) : super(key: key);

  @override
  State<StatefulWidget> createState() {
    return TapPaymentState();
  }
}

class TapPaymentState extends State<TapPayment> {
  late final WebViewController _controller;
  String checkoutUrl = 'https://tap.company';
  String navUrl = 'tap.company';
  bool loading = true;
  bool pageLoading = true;
  bool loadingError = false;
  late TapServices services;
  int pressed = 0;

  loadPayment() async {
    setState(() {
      loading = true;
    });
    try {
      Map getPayment = await services.sendPayment();
      if (getPayment['error'] == false &&
          getPayment['message']?['transaction']?["url"] != null) {
        setState(() {
          checkoutUrl = getPayment['message']['transaction']["url"].toString();
          navUrl = getPayment['message']['transaction']["url"].toString();
          loading = false;
          pageLoading = false;
          loadingError = false;
        });
        _controller.loadRequest(Uri.parse(checkoutUrl));

      } else {
        widget.onError(getPayment);
        setState(() {
          loading = false;
          pageLoading = false;
          loadingError = true;
        });
      }
    } catch (e) {
      widget.onError(e);
      setState(() {
        loading = false;
        pageLoading = false;
        loadingError = true;
      });
    }
  }

  @override
  void initState() {
    super.initState();
    var formData = {};
    //formData = widget.paymentData;
    formData['post'] = {"url": widget.postUrl};
    formData['redirect'] = {"url": widget.redirectUrl};
    services = TapServices(
        apiKey: widget.apiKey,
        paymentData: {...widget.paymentData, ...formData});
    setState(() {
      navUrl = 'checkout.payments.tap.company';
    });
    loadPayment();

    // #docregion platform_features
    late final PlatformWebViewControllerCreationParams params;
    if (WebViewPlatform.instance is WebKitWebViewPlatform) {
      params = WebKitWebViewControllerCreationParams(
        allowsInlineMediaPlayback: true,
        mediaTypesRequiringUserAction: const <PlaybackMediaTypes>{},
      );
    } else {
      params = const PlatformWebViewControllerCreationParams();
    }

    final WebViewController controller =
        WebViewController.fromPlatformCreationParams(params);
    // #enddocregion platform_features

    controller
      ..setJavaScriptMode(JavaScriptMode.unrestricted)
      ..setBackgroundColor(const Color(0x00000000))
      ..setNavigationDelegate(
        NavigationDelegate(
          onProgress: (int progress) {
            debugPrint('WebView is loading (progress : $progress%)');
          },
          onPageStarted: (String url) {
            setState(() {
              pageLoading = true;
              loadingError = false;
            });
            debugPrint('Page started loading: $url');
          },
          onPageFinished: (String url) {
            setState(() {
              navUrl = url;
              pageLoading = false;
            });
          },
          onWebResourceError: (WebResourceError error) {
            debugPrint('''
              Page resource error:
              code: ${error.errorCode}
              description: ${error.description}
              errorType: ${error.errorType}
              isForMainFrame: ${error.isForMainFrame}
          ''');
          },
          onNavigationRequest: (NavigationRequest request) {
            if (request.url.startsWith('https://www.youtube.com/')) {
              debugPrint('blocking navigation to ${request.url}');
              return NavigationDecision.prevent;
            }
            if (request.url.contains(widget.redirectUrl)) {
              final uri = Uri.parse(request.url);
              debugPrint("Got back: ${uri.queryParameters}");
              Navigator.pushReplacement(
                context,
                MaterialPageRoute(
                    builder: (context) => CompletePayment(
                        url: request.url,
                        services: services,
                        onSuccess: widget.onSuccess,
                        onError: widget.onError)),
              );
            }
            debugPrint('allowing navigation to ${request.url}');
            return NavigationDecision.navigate;
          },
          onUrlChange: (UrlChange change) {
            debugPrint('url change to ${change.url}');
          },
        ),
      )
      ..addJavaScriptChannel(
        'Toaster',
        onMessageReceived: (JavaScriptMessage message) {
          ScaffoldMessenger.of(context).showSnackBar(
            SnackBar(content: Text(message.message)),
          );
        },
      );

    // #docregion platform_features
    if (controller.platform is AndroidWebViewController) {
      AndroidWebViewController.enableDebugging(true);
      (controller.platform as AndroidWebViewController)
          .setMediaPlaybackRequiresUserGesture(false);
    }
    // #enddocregion platform_features

    _controller = controller;

    // Enable hybrid composition.
  }

  @override
  Widget build(BuildContext context) {
    return WillPopScope(
      onWillPop: () async {
        if (pressed < 2) {
          setState(() {
            pressed++;
          });
          final snackBar = SnackBar(
              content: Text(
                  'Press back ${3 - pressed} more times to cancel transaction'));
          ScaffoldMessenger.of(context).showSnackBar(snackBar);
          return false;
        } else {
          return true;
        }
      },
      child: Scaffold(
          appBar: AppBar(
            backgroundColor: const Color(0xFF272727),
            leading: GestureDetector(
              child: const Icon(
                Icons.arrow_back_ios,
                color: Colors.white70,
              ),
              onTap: () => Navigator.pop(context),
            ),
            title: Row(
              children: [
                Expanded(
                    child: Container(
                  padding:
                      const EdgeInsets.symmetric(horizontal: 16, vertical: 8),
                  decoration: BoxDecoration(
                      color: Colors.black.withOpacity(0.3),
                      borderRadius: BorderRadius.circular(20)),
                  child: Row(
                    children: [
                      Icon(
                        Icons.lock_outline,
                        color: Uri.parse(navUrl).hasScheme
                            ? Colors.green
                            : Colors.blue,
                        size: 18,
                      ),
                      const SizedBox(width: 5),
                      Expanded(
                        child: Text(
                          navUrl,
                          overflow: TextOverflow.ellipsis,
                          style: const TextStyle(
                              fontSize: 14, color: Colors.white70),
                        ),
                      ),
                      SizedBox(width: pageLoading ? 5 : 0),
                      pageLoading
                          ? const SpinKitFadingCube(
                              color: Color(0xFFEB920D),
                              size: 10.0,
                            )
                          : const SizedBox()
                    ],
                  ),
                ))
              ],
            ),
            elevation: 0,
          ),
          body: SizedBox(
            height: MediaQuery.of(context).size.height,
            width: MediaQuery.of(context).size.width,
            child: loading || (checkoutUrl == 'https://tap.company' && loadingError == false)
                ? const Column(
                    children: [
                      Expanded(
                        child: Center(
                          child: SpinKitFadingCube(
                            color: Color(0xFFEB920D),
                            size: 30.0,
                          ),
                        ),
                      ),
                    ],
                  )
                : loadingError
                    ? Column(
                        children: [
                          Expanded(
                            child: Center(
                              child: NetworkError(
                                  loadData: loadPayment,
                                  message: "Something went wrong,"),
                            ),
                          ),
                        ],
                      )
                    : Column(
                        children: [
                          Expanded(
                            child: WebViewWidget(controller: _controller),
                          ),
                        ],
                      ),
          )),
    );
  }
}
